src-bindings/vscode/vscode.ml

```diff
@@ -2310,6 +2327,32 @@ module TreeItemCollapsibleState = struct
   [@@js.enum] [@@js]
 end
 
+module CustomDocument = struct
+  include Interface.Make ()
+
+  module OnDidChange = Event.Make (Js.Unit)
+  module OnDidDispose = Event.Make (Js.Unit)
+
+  include
+    [%js:
+    val uri : t -> Uri.t [@@js.get]
+
+    val dispose : t -> unit [@@js.call]
+
+    val onDidChange : t -> OnDidChange.t [@@js.get]
+
+    val onDidDispose : t -> OnDidDispose.t [@@js.get]
```

Comment:
I don't think `CustomDocument` has these properties? It's not listed in the api documentation: https://code.visualstudio.com/api/references/vscode-api#CustomDocument

Reply:
Yes, that's correct, and actually we don't even use the `onDidDispose`. I initially didn't include them to closely mirror the Vscode API. My plan was to add them in the `Cm_document` module. It would become:

```diff
diff --git a/src/cm_editor.ml b/src/cm_editor.ml
index 6ba2ced..a9243a3 100644
--- a/src/cm_editor.ml
+++ b/src/cm_editor.ml
@@ -7,8 +7,23 @@ module Cm_document : sig
   val content : t -> (string, string) result Promise.t
 
   val create : uri:Uri.t -> t
+
+  val onDidChange : t -> Js.Unit.t Event.t
 end = struct
   include CustomDocument
+  module OnDidChange = Event.Make (Js.Unit)
+
+  include
+    [%js:
+    val onDidChange : t -> OnDidChange.t [@@js.get]
+
+    val create :
+         uri:Uri.t
+      -> ?onDidChange:OnDidChange.t
+      -> dispose:(unit -> unit)
+      -> unit
+      -> t
+      [@@js.builder]]
 
   let content t =
     let uri = uri t in
@@ -20,8 +35,6 @@ end = struct
     let module EventEmitter = EventEmitter.Make (Js.Unit) in
     let onDidChange_event_emitter = EventEmitter.make () in
     let onDidChange_event = EventEmitter.event onDidChange_event_emitter in
-    let onDidDispose_event_emitter = EventEmitter.make () in
-    let onDidDispose_event = EventEmitter.event onDidDispose_event_emitter in
     let watcher =
       Workspace.createFileSystemWatcher
         (`String (Uri.fsPath uri))
@@ -32,10 +45,8 @@ end = struct
         ~listener:(fun _ -> EventEmitter.fire onDidChange_event_emitter ())
         ()
     in
-    create ~uri ~onDidChange:onDidChange_event ~onDidDispose:onDidDispose_event
-      ~dispose:(fun () ->
-        EventEmitter.fire onDidDispose_event_emitter ();
-        Disposable.dispose disposable)
+    create ~uri ~onDidChange:onDidChange_event
+      ~dispose:(fun () -> Disposable.dispose disposable)
       ()
 end
```

But it means that we would have to preprocess the file with the `gen_js_api` ppx, i.e adding the following to the `src/dune` file:

```bash
 (preprocess
  (pps gen_js_api.ppx))
```

Not sure if that's something we want to avoid or if it's totally ok?

Reply:
I think it's fine to add the `preprocess` field. It's more correct to have `onDidChange` in the `Cm_document` module.

Reply:
Thanks, done in https://github.com/ocamllabs/vscode-ocaml-platform/pull/798/commits/40a34fbec022ee31e1c7aeeee3bf457dd42c1961